### PR TITLE
wayland_spinner

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Re-instate the "spinner" when starting miral-shell (re-implementing it in Wayland)

This builds on two PRs that haven't landed yet:

 o Conceal the full type of the "Splash" from the window management policy (#417)
 o Launcher for internal Wayland clients (#416)

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Fedora says:

/usr/bin/ld: cannot find -lwayland-egl
collect2: error: ld returned 1 exit status

**Comment by AlanGriffiths** [comment]
Hmm, that wasn't it. And /lib64/libwayland-egl.so comes from wayland-devel, so that ought to be enough...

File: examples/miral-shell/spinner/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# -*- Mode: CMake; indent-tabs-mode: nil; tab-width: 2 -*-
+#
+# Copyright © 2014 Canonical Ltd.
```
Comment (gerboland) [inline examples/miral-shell/spinner/CMakeLists.txt]: 2018

File: examples/miral-shell/spinner/eglapp.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright © 2013 Canonical Ltd.
```
Comment (gerboland) [inline examples/miral-shell/spinner/eglapp.h]: -2018? And remove Daniel as author, it's a trivial bit of boilerplate.

File: examples/miral-shell/spinner/eglspinner.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+#include <assert.h>
+#include <glib.h>
+#include <string.h>
+#include <GLES2/gl2.h>
+#include <sys/stat.h>
+#include <signal.h>
+#include <atomic>
+#include <mutex>
+
+//#include <mir_toolkit/mir_client_library.h>
```
Comment (gerboland) [inline examples/miral-shell/spinner/eglspinner.cpp]: just remove

File: examples/miral-shell/spinner/eglspinner.cpp [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+    "}                                                    \n";
+
+//std::atomic<bool> dying{false};
+//
+//void lifecycle_event_callback(MirConnection* /*connection*/, MirLifecycleState state, void* context)
+//{
+//    if (state == mir_lifecycle_connection_lost)
+//        static_cast<decltype(dying)*>(context)->store(true);
+//}
+}
```
Comment (gerboland) [inline examples/miral-shell/spinner/eglspinner.cpp]: Remove?

File: examples/miral-shell/spinner/eglspinner.cpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    GLuint prog[2];
+    GLuint texture[2];
+    GLint vpos[2];
+    GLint theta;
+    GLint fadeGlow;
+    GLint fadeLogo;
+    GLint aTexCoords[2];
+    GLint sampler[2];
+
+//    mir_connection_set_lifecycle_event_callback(connection, &lifecycle_event_callback, &dying);
```
Comment (gerboland) [inline examples/miral-shell/spinner/eglspinner.cpp]: Kept just as reminder to set up equivalent with Wayland?

File: examples/miral-shell/spinner/miregl.cpp [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+
+    if (strcmp(interface, "wl_compositor") == 0)
+    {
+        app->compositor = static_cast<decltype(app->compositor)>(wl_registry_bind(registry, id, &wl_compositor_interface, 3));
+    }
+    else if (strcmp(interface, "wl_shm") == 0)
+    {
+        app->shm = static_cast<decltype(app->shm)>(wl_registry_bind(registry, id, &wl_shm_interface, 1));
+        // Normally we'd add a listener to pick up the supported formats here
+        // As luck would have it, I know that argb8888 is the only format we support :)
```
Comment (gerboland) [inline examples/miral-shell/spinner/miregl.cpp]: for now?

File: examples/miral-shell/spinner/CMakeLists.txt (hunk 1 above)
Comment (AlanGriffiths) [reply examples/miral-shell/spinner/CMakeLists.txt]: Actually, no. This is just a restored file.

File: examples/miral-shell/spinner/eglapp.h (hunk 2 above)
Comment (AlanGriffiths) [reply examples/miral-shell/spinner/eglapp.h]: Actually, no. This is just a restored file.

File: examples/miral-shell/spinner/miregl.cpp (hunk 6 above)
Comment (AlanGriffiths) [reply examples/miral-shell/spinner/miregl.cpp]: That's C&P from examples/mirclient/wayland_client.c

**Comment by Saviq** [comment]
clang complains:
```
CMakeFiles/miral-shell.dir/shell_main.cpp.o: In function `StartupInternalClient<SpinnerSplash>':
./build-amd64/examples/miral-shell/./include/miral/miral/internal_client.h:59: undefined reference to `miral::StartupInternalClient::StartupInternalClient(std::function<void (wl_display*)>, std::function<void (std::weak_ptr<mir::scene::Session>)>)'
```

**Comment by AlanGriffiths** [comment]
Why has clang started complaining?

1. I don't see StartupInternalClient related changes between here and trunk.
2. Building with clang on 18.04 locally works.
3. Doesn't look like CI has an updated clang

**Comment by gerboland** [comment]
LGTM bors r+

**Comment by AlanGriffiths** [comment]
@gerboland bors doesn't notice edits...

bors r=gerboland

**Pull request merged**